### PR TITLE
Finish SQLProduct migration

### PR DESCRIPTION
**Pull request by emord**
Just opening for the sake of public work. builds off of https://github.com/dimagi/commcare-hq/pull/25157/. I think it should be possible to complete this next week. Its been in progress for nearly 5 years (> 20k PRs have passed since it was first introduced) https://github.com/dimagi/commcare-hq/pull/4043

- [x] SQLProuduct support for from_excel
- [x] SQLProduct support archive/unarchive
- [x] save directly to SQLProduct (bypass couch)
- [x] move make_product in tests to SQLProduct only

### CODE REVIEW

**Review by millerdev — COMMENTED** [review]
🏅 Looking great!

**Review by millerdev — COMMENTED** [review]
⬅️ (review progress marker)

File: corehq/apps/products/forms.py [hunk 1]
```diff
@@ -65,8 +65,13 @@ def __init__(self, product, *args, **kwargs):
     def clean_name(self):
         name = self.cleaned_data['name']
 
-        other_products = [p for p in Product.by_domain(self.product.domain) if p._id != self.product._id]
-        if name in [p.name for p in other_products]:
+        num_other_products_with_name = (
+            SQLProduct.objects
+            .filter(domain=self.product.domain, name=name)
+            .exclude(product_id=self.product._id)
+        ).count()
```
[the remark this answers is not in the document]
Comment (emord) [reply corehq/apps/products/forms.py]: Nice. I forgot about `exists` https://github.com/dimagi/commcare-hq/pull/25194/commits/7758abb9f12ff646f17c27e213daa4b009194de8

File: corehq/apps/commtrack/tests/util.py [hunk 2]
```diff
@@ -103,12 +103,13 @@ def bootstrap_location_types(domain):
...
 
 
-def make_product(domain, name, code, program_id):
+def make_product(domain, name, code, program_id=None):
     p = Product()
     p.domain = domain
     p.name = name
     p.code = code.lower()
-    p.program_id = program_id
+    if program_id:
```
[the remark this answers is not in the document]
Comment (emord) [reply corehq/apps/commtrack/tests/util.py]: there are some tests with it `''`and that's fine. i ended up changing this to be nicer anyways https://github.com/dimagi/commcare-hq/pull/25194/commits/64ec64db464bdc8a886f9c46eacac6607e270a7e

File: custom/ewsghana/reports/maps.py [hunk 3]
```diff
@@ -34,6 +34,13 @@ def active_location(self):
             location_id=self.config.get('location_id') or self.default_location_id
         )
 
+    @property
+    @memoized
+    def active_product(self):
+        product_id = self.config.get('product_id')
+        if product_id:
+            return SQLProduct.objects.get(domain=self.domain, product_id=product_id)
```
[the remark this answers is not in the document]
Comment (emord) [reply custom/ewsghana/reports/maps.py]: looks like i accidentally pushed this in a separate commit to use active_objects https://github.com/dimagi/commcare-hq/pull/25194/commits/b6215f78cce866f16ae45019ef7f71172409dfbd#diff-2c50b07a49c843fb5c5fd3eeab0e5b5aR41

**Review by millerdev — APPROVED** [review]

**Review by esoergel — COMMENTED** [review]
This is fantastic!  In defense of whoever created `SQLProduct`, I think it was originally intended as a read-only clone for reporting purposes, similar to how we use ES.  We're much better off strictly on SQL, IMO though, so I'm very happy to see this progress.

File: custom/ewsghana/tests/ews_test_reminders.py [hunk 4]
```diff
@@ -21,12 +22,8 @@ def setUpClass(cls):
     def setUp(self):
         super(EWSTestReminders, self).setUp()
         self.facility = make_loc('test-faciity', 'Test Facility', TEST_DOMAIN, 'Polyclinic')
-        self.commodity = Product(domain=TEST_DOMAIN, name='Drug A', code_='ab', unit='cycle')
-        self.commodity.save()
-
-        self.commodity2 = Product(domain=TEST_DOMAIN, name='Drug B', code_='cd', unit='cycle')
-        self.commodity2.save()
-
+        self.commodity = make_product(TEST_DOMAIN, 'Drug A', '_ab', unit='cycle')
```
[the remark this answers is not in the document]
Comment (emord) [reply custom/ewsghana/tests/ews_test_reminders.py]: No. it also caused test errors https://github.com/dimagi/commcare-hq/pull/25194/commits/e33e2bbb38d70edca0abe4be42b8557ee875e0c6

**Comment by emord** [comment]
> In defense of whoever created SQLProduct, I think it was originally intended as a read-only clone for reporting purpose

Yeah that's been pretty clear on these last few steps to get it over the hump.

Anyways, I believe that this is ~done functionality wise. the only remaining reference to `Product` is in the `clone_domain` mgmt command (I thought it wouldn't even work anymore, but there are tests around it so that makes think it does work). I don't totally trust our test suite to test the import and editing/creation pages, so I'll at least do some manual tests on those.

This is a big enough change that I'd rather do it after I get back from PTO on the 23rd unless someone else particularly wants to see it through

**Review by millerdev — APPROVED** [review]

**Comment by emord** [comment]
had some errors from ewsghana tests, but those are now removed, so I merged master. Will look to have this done with tests passing by end of week (or at least pull out the easy test changes into a separate PR to reduce the size of this one)

**Review by millerdev — APPROVED** [review]

**Comment by emord** [comment]
definitely regret not pulling these out into separate PRs earlier (before ewsghana was removed). Will plan to release next week after thanksgiving since i'm offline most of hte week

File: corehq/apps/products/models.py [hunk 5]
```diff
@@ -219,123 +154,49 @@ def from_excel(cls, row, custom_data_validator):
         id = row.get('id')
         if id:
             try:
-                p = cls.get(id)
+                product = cls.objects.get(location_id=id)
```
Comment (snopoke) [inline corehq/apps/products/models.py]: shouldn't this be `product_id=id`?

**Review by snopoke — COMMENTED** [review]
One question but otherwise looks good

**Comment by emord** [comment]
I tested this out on staging for a bit. Most things appeared to work at first blush, but the errors I ran into made me super cautious about this (particularly the late one where a form used the [hidden `_doc` to get a dict](https://github.com/dimagi/commcare-hq/pull/25194/commits/9060b0e8ba8e055e2b2a1698af8e14abc04cb1e1) )

The things I found that are definitely untested are:
* the forms (both updated and created)
* exports

Those should probably get tests added before a final migration to ensure that the various private interfaces used here are not going to cause issues

**Pull request closed without merging**